control_software/scripts/hera_fem_switch_loop.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
+# -*- mode: python; coding: utf-8 -*-
+# Copyright 2017-2019 the HERA Collaboration
+# Licensed under the 2-clause BSD license.
+
+"""Loop through all FEM states for given SNAPs and antenna indices."""
+
+import os
+import sys
+import numpy as np
+import argparse
+import logging
+from hera_corr_f import HeraCorrelator, SnapFengine, helpers
+
+if sys.version_info[0] < 3:
+    # py2
+    from plotly import graph_objs as go, subplots, io as pio
+else:
+    # py3
+    from plotly import graph_objects as go, subplots, io as pio
+
+logger = helpers.add_default_log_handlers(logging.getLogger(__name__),
+                                          fglevel=logging.NOTSET)
+
+
+def main(redishost='redishost', hostname=None, antenna_input=None,
+         integration_time=None, do_not_initialize=False):
+    if hostname is None:
+        raise ValueError("Must specify a hostname or list of hostames. "
+                         "To run on all Snaps set hostame='all'")
+    # not actually using the connection
+    # but will bw used tempo manipulate the monitoring state
+    # and find all hosts if needed.
+    c = HeraCorrelator(redishost=redishost,
+                       passive=True)
+    if hostname:
+        if not isinstance(hostname, (list, tuple)):
+            hostname = [hostname]
+
+    hostname = sorted(hostname)
+
+    if hostname == ['all'] and not antenna_input:
+        logger.info("Using all SNAPs from redis configuration")
+        SNAP_HOST = sorted(list(c.config['fengines'].keys()))
+        ANTENNA_INPUT = np.arange(0, 6)
+    elif hostname != ['all'] and not antenna_input:
+        SNAP_HOST = hostname
+        ANTENNA_INPUT = np.arange(0, 6)
+    else:
+        SNAP_HOST = hostname
+        ANTENNA_INPUT = antenna_input
+
+    correlations = {}
+    for host in SNAP_HOST:
+        # Disable the monitoring process for 5 minutes,
+        # to stop it interfering with this script
+        c.disable_monitoring(300, wait=True)
+
+        snap = SnapFengine(host, redishost=redishost)
+
+        if not snap.is_programmed():
```

Comment:
if you need to program the board, you'll need to initialize.

Reply:
Just want to make sure I understand, must not have caught on the first time we talked about the snaps. Something like the following?
```
if not programmed:
    unless do not init:
        snap.initialize()
    snap.fpga.transport.prog_user_image()
```

Reply:
been a long day, but the logic should be:
```
if not programmed:
     program()
    initialize()
else:
    if not do_not_init:
        initialize()
```

I'll let you equate the boolean logic :)

Basically, if you call prog_user_image() you must also call initialize. Because programming the board will bring it up in an unconfigured state

Reply:
Thanks for the clarification.